python/ray/dag/compiled_dag_node.py

```diff
@@ -496,6 +603,145 @@ def exec_operation(
             return self._write()
 
+    def set_stream_buffer(self, op: _DAGNodeOperation, data: Any):
+        """
+        Store the intermediate result of a READ or COMPUTE operation
+        in the stream buffer.
+
+        Args:
+            op: The operation that generates the intermediate result.
+            data: The intermediate result of a READ or COMPUTE operation.
+        """
+        self._stream_buffer[op.next_operation()] = data
+
+    def reset_stream_buffer(self, op: _DAGNodeOperation) -> Any:
+        """
+        Retrieve the intermediate result of a READ or COMPUTE operation
+        from the stream buffer and clear the entry from the buffer.
+
+        Returns:
+            The intermediate result of a READ or COMPUTE operation.
+        """
+        return self._stream_buffer.pop(op)
+
+    def _stream_read(self, op: _DAGNodeOperation) -> bool:
+        """
+        Read input data from upstream DAG nodes and cache the intermediate result.
+
+        Args:
+            op: The READ operation.
+
+        Returns:
+            True if system error occurs and exit the loop; otherwise, False.
+        """
+        exit = False
+        try:
+            input_data = self.input_reader.read()
+            self.set_stream_buffer(op, input_data)
+        except RayChannelError:
+            # Channel closed. Exit the loop.
+            exit = True
+        return exit
+
+    def _stream_compute(self, op: _DAGNodeOperation, exec_stream, class_handle) -> bool:
+        """
+        Retrieve the intermediate result from the READ operation and perform the
+        computation on the provided execution stream. Then, cache the new
+        intermediate result.
+
+        Args:
+            op: The compute operation.
+            exec_stream: The CUDA stream to execute the compute operation.
+            class_handle: An instance of the class to which the actor belongs. For
+                example, the type of `class_handle` is <class 'xxxx.Worker'> if the
+                actor belongs to the `class Worker` class.
+
+        Returns:
+            True if system error occurs and exit the loop; otherwise, False.
+        """
+        input_data = self.reset_stream_buffer(op)
+        method = getattr(class_handle, self.method_name)
+        try:
+            _process_return_vals(input_data, return_single_output=False)
+        except Exception as exc:
+            # Previous task raised an application-level exception.
+            # Propagate it and skip the actual task. We don't need to wrap the
+            # exception in a RayTaskError here because it has already been wrapped
+            # by the previous task.
+            self.set_stream_buffer(exc)
+            return False
+
+        channel_results = []
+        for entry in input_data:
+            if isinstance(entry, tuple):
+                channel_result, event = entry
+                if event:
+                    event.synchronize()
```

Comment:
Note that this will also block the CPU thread, not just the CUDA stream that needs to read the data. You probably want to just have the CUDA stream [wait on the event](https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.Stream.html#cupy.cuda.Stream.wait_event).

Reply:
I think that's what we want? We don't know whether user UDF runs on GPU or CPU, so if there is a data dependency, we need to sync all?

Reply:
If there is a GPU -> CPU dependency, then there's going to be some CUDA memcpy to move the data from GPU -> CPU. So we don't need to block the entire CPU task on the GPU task, we just need to make sure the memcpy executes after the GPU task (which it should as long as it's executed on the correct stream).

Reply:
Just to clarify: this event is a recv event on a recv stream, and here the compute operation runs on CPU or execute stream of the GPU. Suppose it runs on CPU, so you are saying as long as the execution operation happens after the "async" recv operation is launched (which it should because they have the same task_index), even without the manual sync with CPU, the memcpy would happen and happens after the recv finishes?

Reply:
Ah not quite. What I was trying to say is that even if the task is executing on CPU, if it depends on the GPU data, then it needs to run some memcpy to actually read the GPU data. The memcpy will execute on `exec_stream`, so we just need to make sure to sync exec_stream with the recv stream (using `stream.wait_event`). We don't need to sync everything including the CPU with the recv stream (using `event.synchronize`).

Reply:
(and if the CPU task does not actually read the GPU data, then it doesn't matter if we sync at all)

Reply:
Make sense, thanks for the clarification!

Actually, cupy.cuda.Event.synchronize() does not block CPU thread when it is created with `block=False` (default param): https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.Event.html#cupy.cuda.Event.synchronize

But it mentioned "Synchronizes all device work to the event", which probably means sync with all streams -- this would sync with the `send` stream. So I will change to use https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.Stream.html#cupy.cuda.Stream.wait_event